cli/src/action/login/oauth/github.rs

```diff
@@ -0,0 +1,238 @@
+            .exchange_code(AuthorizationCode::new(auth_code))
+            .set_pkce_verifier(pkce_verifier)
+            .request(|req| {
+                Client::new()
+                    .execute(to_reqwest_req(req)?)
+                    .map(to_oauth2_res)
+                    .map_err(|_err| {
+                        OAuth2ProviderError("Unable to execute for OAuth2 Access Token".into())
+                    })
+            })
+            .map_err(|_err| OAuth2ProviderError("Unable to request OAuth2 Access Token".into()))?;
+
+        Ok(UserTokens {
+            provider_type: ProviderType::Github,
+            username: Self::username(client_id, client_secret, token_result.access_token())?,
+            access_token: token_result.access_token().secret().clone(),
+            refresh_token: None,
+        })
+    }
+}
+
+impl GithubProvider {
+    #[inline]
+    pub(super) const fn new(
+        client_id: Option<&'static str>,
+        client_secret: Option<&'static str>,
+    ) -> Self {
+        Self {
+            client_id,
+            client_secret,
+        }
+    }
+
+    fn username(
+        client_id: &str,
+        client_secret: &str,
+        access_token: &AccessToken,
+    ) -> Result<String, OAuth2ProviderError> {
+        let body = serde_json::to_string(&GithubAccesToken {
+            access_token: access_token.secret(),
+        })
+        .map_err(|_| OAuth2ProviderError("Unable to write token request to JSON".into()))?;
+
+        let response = reqwest::blocking::Client::new()
+            .post(&format!(
+                "https://api.github.com/applications/{}/token",
+                client_id
+            ))
+            .header("User-Agent", "Splinter CLI")
+            .basic_auth(client_id, Some(client_secret))
+            .body(body)
+            .send();
+
+        match response {
+            Ok(res) if res.status().is_success() => {
+                let token_info: GithubTokenInfo = res
+                    .json()
+                    .map_err(|_err| OAuth2ProviderError("Unable to fetch username".into()))?;
+
+                Ok(token_info.user.login)
+            }
+            Ok(res) => {
+                debug!(
+                    "Received response {}: {}",
+                    res.status(),
+                    res.text()
+                        .unwrap_or_else(|_| "<empty or binary content>".into())
+                );
+                Err(OAuth2ProviderError(
+                    "Unable to validate OAuth1 access token.".into(),
+                ))
+            }
+            Err(err) => {
+                debug!("Unable to validate access token due to error: {}", err);
+                Err(OAuth2ProviderError(
+                    "Unable to validate OAuth1 access token.".into(),
+                ))
+            }
+        }
+    }
+}
+
+#[derive(serde::Serialize)]
+struct GithubAccesToken<'a> {
+    access_token: &'a str,
+}
+
+#[derive(serde::Deserialize)]
+struct GithubTokenInfo {
```

Comment:
Should probably be `GithubUserInfo`?

Reply:
Technically, this is the response from the gitub token endpoint, so this is accurately named.  It will also be expanded on when refresh is implemented.